### DISCUSSION TITLE
Optimizing a module with several signatures

### DISCUSSION THREAD

**Discussion by dampeel365** [stanfordnlp/dspy#1670] [Q&A]
Hello,

Optimizing a Module holding different signatures (for instance 2 CoT) chaining seems functionnal but, the generated json model looks weird, especially for the not-augmented samples which doesn't seems to stick to input/output of the signature but looks like a global intput/output for the whole module....

Is it supposed to work that way ?
Is it recommended to train such modules this way ? Or would you suggest to train each signature in a different module then to gather them then ?

**Comment by aniruddhaadak80** [comment]
From my point of view, the weird JSON is a clue that the optimizer may be reasoning at the module boundary while your expectations are tied to each individual signature. That mismatch often shows up first in synthesized training artifacts before it shows up in runtime quality.

If the signatures play very different roles, I would be inclined to optimize the submodules separately first and then compare that against whole-module optimization rather than assume one composite pass will preserve each boundary cleanly.
